editor/emacs/config.org
#+TITLE: My Emacs Config (omg)
#+AUTHOR: Sean Chalmers
#+EMAIL: sclhiannan@gmail.com
#+OPTIONS: num:nil

* Sanity settings
  Disable the comp warning spam
  #+begin_src emacs-lisp
(setq comp-async-compilation t)
(setq comp-async-env-modifier-form
      '((setenv "LIBRARY_PATH"
                (concat
                 (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in with pkgs; stdenv.lib.makeLibraryPath [stdenv.cc.cc stdenv.glibc])'")
                 ":"
                 (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in with pkgs; lib.getLib libgccjit + /lib/gcc/x86_64-unknown-linux-gnu/9.3.0 )'")))))
(setq-default comp-async-report-warnings-errors nil)
  #+end_src
  Disable tabs, we're not monsters
  #+BEGIN_SRC emacs-lisp
;; (setq debug-on-error t)
(setq-default indent-tabs-mode nil)
(setq-default electric-indent-inhibit t)
  #+END_SRC

  Enable the unicodening to annoy jackd
  #+begin_src emacs-lisp
(set-language-environment "UTF-8")
(set-default-coding-systems 'utf-8)
  #+end_src

  Set a global tab width to two spaces
  #+BEGIN_SRC emacs-lisp
(setq-default tab-width 2)
  #+END_SRC

  Fix up some modes that have their own snowflake variables for indentation.
  #+BEGIN_SRC emacs-lisp
(setq-default css-indent-offset 2)
  #+END_SRC

  Move our backups out of sight
  #+BEGIN_SRC emacs-lisp
(setq backup-directory-alist `((".*" . ,temporary-file-directory)))
(setq auto-save-file-name-transforms `((".*" ,temporary-file-directory t)))
  #+END_SRC
  Set out GC threshold to 50Mb instead of 800Kb
  #+BEGIN_SRC emacs-lisp
(setq gc-cons-threshold 50000000)
  #+END_SRC

  #+BEGIN_SRC emacs-lisp
    ;; Enable line highlight mode, because I find it useful.
    (global-hl-line-mode t)
    ;; Column numbers too
    (column-number-mode t)
    ;; Default fill-width
    (setq-default fill-column 100)
    ;; Show trailing whitespace
    (setq-default show-trailing-whitespace t)
    ;; Hide the scrollbar
    (toggle-scroll-bar -1)
  #+END_SRC
* Purdy
  Add some themey goodness.
  #+BEGIN_SRC emacs-lisp
    (use-package monokai-theme :ensure t)
    (load-theme 'monokai t)
    ;; (load-theme 'wombat t)
    ;; (load-theme 'github-modern t)
  #+END_SRC
* Diminish and Delight
** Diminish
   #+BEGIN_SRC emacs-lisp
     (use-package diminish
       :ensure t
       :config
       (diminish 'undo-tree-mode "UT")
       )
   #+END_SRC
** Delight
   #+BEGIN_SRC emacs-lisp
     (use-package delight
       :ensure t
       )
   #+END_SRC
* =which-key= for discoverability
  #+BEGIN_SRC emacs-lisp
    (setq which-key-enable-extended-define-key t)
    (use-package which-key
      :ensure t
      :diminish which-key-mode
      :init
      (setq which-key-separator " ")
      (setq which-key-prefix-prefix "+")
      :config
      (which-key-mode 1)
      )

    (defun manky/add-leader-prefix-name (k n)
      "Using the current evil-leader key, define a name for which-key to display"
      (which-key-add-key-based-replacements (concat (symbol-value 'evil-leader/leader) " " k) n))
  #+END_SRC

* Evil mode for great justice
  With hopefully less RSI..

  #+BEGIN_SRC emacs-lisp
    (use-package evil
      :ensure t
      :config
      (use-package evil-leader
        :ensure t
        :config
        (global-evil-leader-mode 1)
        (evil-leader/set-leader "<SPC>"))

      (use-package evil-surround
        :ensure t
        :config (global-evil-surround-mode))

      (use-package evil-escape
        :ensure t
        :diminish evil-escape-mode
        :init (setq-default evil-escape-key-sequence "jk")
        :config (evil-escape-mode 1))

      (use-package evil-nerd-commenter
        :ensure t
        :config
        (manky/add-leader-prefix-name "c" "comments")
        (evil-leader/set-key
          "cl" 'evilnc-comment-or-uncomment-lines
          "cc" 'evilnc-copy-and-comment-lines
          "cp" 'evilnc-comment-or-uncomment-paragraphs
          "cr" 'comment-or-uncomment-region
          "cv" 'evilnc-toggle-invert-comment-line-by-line
          "."  'evilnc-copy-and-comment-operator
          "\\" 'evilnc-comment-operator ; if you prefer backslash key
          )
        )
      ;; (use-package evil-magit
      ;;   ;; :ensure t
      ;;   :after (evil magit)
      ;;   )
      (use-package powerline :ensure t)
      (use-package powerline-evil
        :ensure t
        :after (powerline)
        :config
        (powerline-evil-vim-color-theme))
      (evil-mode 1)
      )
  #+END_SRC

* Avy for jumping
  #+BEGIN_SRC emacs-lisp
    (use-package avy
      :ensure t
      :config
      (manky/add-leader-prefix-name "j" "avy")
      (evil-leader/set-key
        "j c" 'avy-goto-char
        "j c" 'avy-goto-char
        "j C" 'avy-goto-char-2
        "j t" 'avy-goto-char-timer
        "j l" 'avy-goto-line
        "j w" 'avy-goto-word-1
        "j W" 'avy-goto-word-0
        "j o" 'avy-org-goto-heading-timer
        "j R" 'avy-org-refile-as-child
        "j b" 'avy-pop-mark
        )
      )
  #+END_SRC
* Clean up the Emacs UI
  #+BEGIN_SRC emacs-lisp
    (scroll-bar-mode -1)
    (tool-bar-mode   -1)
    (tooltip-mode    -1)
    (menu-bar-mode   -1)
    (setq-default inhibit-startup-screen t)
    ;; (setq default-frame-alist '((font . "-ADBO-Source Code Pro-normal-normal-normal-*-14-*-*-*-m-0-iso10646-1")))
    ;; (setq default-frame-alist '((font . "-POOP-Fixedsys Excelsior 3.01-normal-normal-normal-*-16-*-*-*-*-0-iso10646-1")))
    ;; (setq default-frame-alist '((font . "-V.R.-PxPlus IBM VGA9-normal-normal-normal-*-15-*-*-*-m-0-iso10646-1")))
    (setq default-frame-alist '((font . "-BE5N-Iosevka-semibold-normal-normal-*-13-*-*-*-d-0-iso10646-1")))
    ;; (setq default-frame-alist '((font . "-UKWN-mononoki-normal-normal-normal-*-13-*-*-*-*-0-iso10646-1")))
    ;; (setq default-frame-alist '((font . "-CTDB-Fira Code-normal-normal-normal-*-13-*-*-*-m-0-iso10646-1")))
    ;; (setq default-frame-alist '((font . "-ADBO-Hasklig-normal-normal-normal-*-12-*-*-*-m-0-iso10646-1")))

  #+END_SRC
* Prescient
  #+begin_src emacs-lisp
  (use-package prescient :ensure t)
  (use-package selectrum-prescient :ensure t :after (prescient))
  (use-package company-prescient :ensure t :after (prescient))
  #+end_src
* Selectrum
  #+begin_src emacs-lisp
    (use-package selectrum
      :ensure t
      :after (selectrum-prescient)
      :config
      ;; to make sorting and filtering more intelligent
      (selectrum-prescient-mode +1)
      (selectrum-mode +1)

      (defvar selectrum--toggle-project-data+ nil)

      (defun selectrum-toggle-project-file-scope+ ()
        "Toggle to project scope when reading file names.
      Depends on `projectile'."
        (interactive)
        (unless minibuffer-completing-file-name
          (user-error "Not reading file names"))
        (require 'projectile)
        (setq selectrum--previous-input-string nil)
        (cond ((and selectrum--toggle-project-data+
                    (string-match "in project: \\'"
                                  (buffer-substring
                                   (point-min) (minibuffer-prompt-end))))
               (let ((inhibit-read-only t))
                 (save-excursion
                   (goto-char (minibuffer-prompt-end))
                   (search-backward " in project")
                   (delete-region (match-beginning 0)
                                  (match-end 0)))
                 (delete-minibuffer-contents))
               (insert (car selectrum--toggle-project-data+))
               (setq selectrum--preprocessed-candidates
                     (cdr selectrum--toggle-project-data+))
               (setq selectrum--toggle-project-data+ nil))
              (t
               (if-let ((input (selectrum-get-current-input))
                        (project (projectile-project-root
                                  (file-name-directory input))))
                   (let* ((inhibit-read-only t)
                          (ematch (file-name-nondirectory input))
                          (cands
                           (mapcar
                            (lambda (i)
                              (add-text-properties
                               0 (length i)
                               `(selectrum-candidate-full
                                 ,(concat project i))
                               i)
                              i)
                            (projectile-project-files project))))
                     (save-excursion
                       (goto-char (minibuffer-prompt-end))
                       (search-backward ":")
                       (insert
                        (apply #'propertize
                               " in project"
                               (text-properties-at (point)))))
                     (setq selectrum--toggle-project-data+
                           (cons
                            input
                            selectrum--preprocessed-candidates))
                     (delete-minibuffer-contents)
                     (insert
                      (concat (abbreviate-file-name project) ematch))
                     (setq selectrum--preprocessed-candidates
                           (lambda (input)
                             (let ((ematch (file-name-nondirectory input)))
                               `((input . ,ematch)
                                 (candidates . ,cands))))))
                 (user-error "Not in project")))))


    ;; (push (cons "C-," 'selectrum-toggle-project-file-scope+) selectrum-minibuffer-bindings)
  )
  #+end_src
* Consult
  #+begin_src emacs-lisp
;; Example configuration for Consult
(use-package consult
  :ensure t
  ;; Replace bindings. Lazily loaded due by `use-package'.
  :bind (("C-x M-:" . consult-complex-command)
         ("C-c h" . consult-history)
         ("C-c m" . consult-mode-command)
         ("C-c k" . consult-keep-lines)
         ("C-c C-k" . consult-hide-lines)
         ("C-x b" . consult-buffer)
         ("C-x 4 b" . consult-buffer-other-window)
         ("C-x 5 b" . consult-buffer-other-frame)
         ("C-x r x" . consult-register)
         ("C-x r b" . consult-bookmark)
         ("M-g g" . consult-goto-line)
         ("M-g M-g" . consult-goto-line)
         ("M-g o" . consult-outline)       ;; "M-s o" is a good alternative.
         ("M-g l" . consult-line)          ;; "M-s l" is a good alternative.
         ("M-g m" . consult-mark)          ;; I recommend to bind Consult navigation
         ("M-g k" . consult-global-mark)   ;; commands under the "M-g" prefix.
         ("M-g r" . consult-ripgrep)       ;; or consult-grep, consult-ripgrep, or consult-git-grep
         ("M-g f" . consult-find)          ;; or consult-locate, my-fdfind
         ("M-g i" . consult-project-imenu) ;; or consult-imenu
         ("M-g e" . consult-error)
         ("M-s m" . consult-multi-occur)
         ("M-y" . consult-yank-pop)
         ("<help> a" . consult-apropos))

  ;; The :init configuration is always executed (Not lazy!)
  :init
  ;; Custom command wrappers. It is generally encouraged to write your own
  ;; commands based on the Consult commands. Some commands have arguments which
  ;; allow tweaking. Furthermore global configuration variables can be set
  ;; locally in a let-binding.
  (defun my-fdfind (&optional dir)
    (interactive "P")
    (let ((consult-find-command "fdfind --color=never --full-path"))
      (consult-find dir)))

  ;; Replace `multi-occur' with `consult-multi-occur', which is a drop-in replacement.
  (fset 'multi-occur #'consult-multi-occur)

  ;; Configure register preview function.
  ;; This gives a consistent display for both `consult-register' and
  ;; the register preview when editing registers.
  (setq register-preview-delay 0
        register-preview-function #'consult-register-preview)

  ;; Configure other variables and modes in the :config section, after lazily loading the package
  :config

  ;; Configure preview. Note that the preview-key can also be configured on a
  ;; per-command basis via `consult-config'.
  ;; The default value is 'any, such that any key triggers the preview.
  ;; (setq consult-preview-key 'any)
  ;; (setq consult-preview-key (kbd "M-p"))
  ;; (setq consult-preview-key (list (kbd "<S-down>") (kbd "<S-up>")))

  ;; Optionally configure narrowing key.
  ;; Both < and C-+ work reasonably well.
  (setq consult-narrow-key "<") ;; (kbd "C-+")
  ;; Optionally make narrowing help available in the minibuffer.
  ;; Probably not needed if you are using which-key.
  ;; (define-key consult-narrow-map (vconcat consult-narrow-key "?") #'consult-narrow-help)

  ;; Optional configure a view library to be used by `consult-buffer'.
  ;; The view library must provide two functions, one to open the view by name,
  ;; and one function which must return a list of views as strings.
  ;; Example: https://github.com/minad/bookmark-view/
  ;; (setq consult-view-open-function #'bookmark-jump
  ;;       consult-view-list-function #'bookmark-view-names)

  ;; Optionally configure a function which returns the project root directory
  (autoload 'projectile-project-root "projectile")
  (setq consult-project-root-function #'projectile-project-root))
  #+end_src
* Async?!
  #+BEGIN_SRC emacs-lisp
    (use-package async
      :ensure t
      :config
      (dired-async-mode 1)
      )
  #+END_SRC
* Popups
  #+BEGIN_SRC emacs-lisp
    (use-package popup
      :ensure t
      )
  #+END_SRC
* Treemacs
** Treemacs main package
   #+BEGIN_SRC emacs-lisp
     (use-package treemacs
       :ensure t
       :config
       (defun treemacs-ignore-flymake (file _)
         (string-match-p (regexp-quote "_flymake\..+") file))
       (push #'treemacs-ignore-flymake treemacs-ignored-file-predicates)
       (treemacs-follow-mode))
   #+END_SRC
** Treemacs evil
   #+BEGIN_SRC emacs-lisp
     (use-package treemacs-evil
       :ensure t
       :after (treemacs evil)
       )
   #+END_SRC
** Treemacs projectile
   #+BEGIN_SRC emacs-lisp
     ;; (use-package treemacs-projectile
     ;;   ;; :ensure t
     ;;   :after (treemacs projectile)
     ;;   :config
     ;;   (evil-leader/set-key
     ;;     "pt" 'treemacs-add-and-display-project
     ;;   )
     ;; )
   #+END_SRC
* Display Line Numbers
  #+BEGIN_SRC emacs-lisp
    (use-package display-line-numbers
      :ensure t
      :config
      (defun display-line-numbers--turn-on ()
        "turn on line numbers but excempting certain major modes defined in `display-line-numbers-exempt-modes'"
        (if (and
             (not (member major-mode '(treemacs)))
             (not (minibufferp)))
            (display-line-numbers-mode)))
      (global-display-line-numbers-mode)
      )
  #+END_SRC
* IEdit for many edited justices
  #+BEGIN_SRC emacs-lisp
    (use-package iedit
      :ensure t
      :bind (("C-;" . iedit-mode))
      )
  #+END_SRC
* Projectile for project goodness
** Projectile
   #+BEGIN_SRC emacs-lisp
     (use-package projectile
       :ensure t
       :delight '(:eval (concat " " (projectile-project-name)))
       :init
       (setq projectile-require-project-root nil)
       :config
       ;; (define-key projectile-mode-map (kbd "s-p") 'projectile-command-map)
       ;; (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)
       (setq projectile-project-search-path '("~/repos"))
       (projectile-mode +1)
       (evil-leader/set-key
         "p" 'projectile-command-map
         )
       )
   #+END_SRC
* Minor Text/Layout utils
** aggressive-indent
   Not in use at the moment
   #+BEGIN_SRC emacs-lisp
     (use-package aggressive-indent
       :ensure t
       :config
       (evil-leader/set-key
         "t a" 'aggressive-indent-mode
         )
       )
   #+END_SRC

** rainbow-delimiters
   #+BEGIN_SRC emacs-lisp
     (use-package rainbow-delimiters
       :ensure t
       ;; There is no global mode, so...
       :hook (prog-mode-hook . rainbow-delimiters-mode)
       )
   #+END_SRC
** smartparens-config
   #+BEGIN_SRC emacs-lisp
     (use-package smartparens
       :ensure t
       :diminish (smartparens-mode . "()")
       :config
       (require 'smartparens-config)
       (smartparens-global-mode t)
       (show-paren-mode t)
       )
   #+END_SRC

* Git!
  #+BEGIN_SRC emacs-lisp
    (use-package magit
      :ensure t
      :diminish magit-auto-revert-mode
      :config
      ;; (global-set-key (kbd "C-x g") 'magit-status)
      (manky/add-leader-prefix-name "g" "git")
      (evil-leader/set-key
        "g s" 'magit-status)
      )
    (use-package forge
      :ensure t
      :after magit
      )
  #+END_SRC
* Direnv
  #+BEGIN_SRC emacs-lisp
    (use-package direnv
      :ensure t
      :config
      (direnv-mode))
  #+END_SRC
* Emmet for xml laziness
  Emmet coding is a life saver when you just have to write XML type things.
  #+BEGIN_SRC emacs-lisp
    (use-package emmet-mode
      :ensure t
      :init
      (add-hook 'sgml-mode-hook 'emmet-mode) ;; Autostart on markup modes
      (add-hook 'css-mode-hook 'emmet-mode) ;; Emmet has CSS prefix helpers
      (setq emmet-move-cursor-between-quotes t) ;; Move to between the inserted tags

      ;; Not sure if I need this one yet, but I'll know it when I hit it
      ;; (setq emmet-self-closing-tag-style " /") ;; default "/"
      ;; only " /", "/" and "" are valid.
      ;; eg. <meta />, <meta/>, <meta>
      )
  #+END_SRC

* Nix/OS integration & tools
** Nix file mode
   Gotta get that highlighting...
   #+BEGIN_SRC emacs-lisp
     (use-package nix-mode
       :ensure t
       :mode ("\\.nix\\'" . 'nix-mode)
       :init
       (defun manky/nix-indent ()
         (make-local-variable 'indent-line-function)
         (setq indent-line-function 'nix-indent-line)
         (setq nix-indent-function 'nix-indent-line)
         )

       (add-hook 'nix-mode-hook 'manky/nix-indent)
       )
   #+END_SRC
** Nix sandbox
   #+BEGIN_SRC emacs-lisp
     (use-package nix-sandbox
       :ensure t
       :after nix-mode
       )
   #+END_SRC
** Universal CTags
   #+begin_src emacs-lisp
(use-package counsel-etags
  :ensure t
  :after (counsel)
  :bind (("C-]" . counsel-etags-find-tag-at-point))
  :init
  (add-hook 'prog-mode-hook
        (lambda ()
          (add-hook 'after-save-hook
            'counsel-etags-virtual-update-tags 'append 'local)))
  :config
  (setq counsel-etags-update-interval 60)
  (push "build" counsel-etags-ignore-directories))
   #+end_src
* Language Modes!! OMG
** Purescript
#+begin_src emacs-lisp
(use-package purescript-mode
  :ensure t
)
(use-package psc-ide
  :ensure t
  :after (purescript-mode)
  :hook (purescript-mode . (lambda ()
           (psc-ide-mode)
               (company-mode)
               (flycheck-mode)
               (turn-on-purescript-indentation)))
)
#+end_src
** CPP
   #+begin_src emacs-lisp
     (use-package clang-format
       :ensure t
       :config
       (evil-leader/set-key "b F" 'clang-format-region)
     )

     (use-package ccls
       :after (lsp-mode)
       :ensure t
       :hook ((c-mode c++-mode objc-mode cuda-mode) . (lambda () (require 'ccls) (lsp)))
       :config
       (setq ccls-executable "ccls")
     )
   #+end_src
** CMake
   #+begin_src emacs-lisp
(use-package cmake-mode :ensure t)
   #+end_src

** OCaml
 #+begin_src emacs-lisp
   (use-package tuareg
     :ensure t
     :load-path "~/.nix-profile/share/emacs/site-lisp"
   )
   (use-package merlin
     :ensure t
     :init
     (add-to-list 'load-path "~/.nix-profile/share/emacs/site-lisp")
     (setq merlin-command "ocamlmerlin")
     :config
     (add-hook 'tuareg-mode-hook 'merlin-mode t)
   )
   (use-package ocp-indent :ensure t)
   #+end_src
** Elixir
 #+BEGIN_SRC emacs-lisp

   (use-package mix :ensure t)
   (use-package elixir-mode
     :ensure t
     :init
     ;; (add-hook 'elixir-mode-hook 'mix-minor-mode)
     )

 #+END_SRC

** Rakudo!!
#+BEGIN_SRC emacs-lisp
(use-package raku-mode
  :ensure t
  :defer t
)
#+END_SRC
** Crystal
#+BEGIN_SRC emacs-lisp
(use-package crystal-mode
:ensure t
)
#+END_SRC
** Haskell
#+BEGIN_SRC emacs-lisp
(use-package haskell-mode
:ensure t
:after flycheck
:config
;; Configure haskell-mode to use cabal new-style builds
(setq haskell-process-type 'cabal-new-repl)
;; Tell company-mode to ask lsp for completions
(setq haskell-completion-backend 'lsp)
;; Make sure we try to use the current nix env if we have one
(setq haskell-process-wrapper-function
  (lambda (args) (apply 'nix-shell-command (nix-current-sandbox) args)))

(setq haskell-hoogle-url "http://localhost:8080/?hoogle=%s")
(evil-leader/set-key
  "h h" 'haskell-hoogle
  )

;; Disable the haskell-stack-ghc checker
(add-to-list 'flycheck-disabled-checkers 'haskell-stack-ghc)
(add-hook 'hack-local-variables-hook #'manky/set-dante-locals nil 'local)

(add-hook 'haskell-mode-hook 'prettify-symbols-mode)
(add-hook 'haskell-mode-hook
  (lambda ()
    (setq tab-width 2)
          (set (make-local-variable 'company-backends)
               (append '((company-capf company-dabbrev-code))
                       company-backends))))

)
#+END_SRC
*** Ormolu
#+BEGIN_SRC emacs-lisp
;; Needed for ormolu package integration.
(use-package reformatter :ensure t)

(use-package ormolu
 :ensure t
 :bind (:map haskell-mode-map ("C-c r" . ormolu-format-buffer)))
#+END_SRC
** JSON
*heavy sigh*
#+BEGIN_SRC emacs-lisp
(use-package json-mode :ensure t)
#+END_SRC
** CSS
#+BEGIN_SRC emacs-lisp
(use-package css-mode :ensure t)
#+END_SRC
** Markdown
#+BEGIN_SRC emacs-lisp
(use-package markdown-mode :ensure t)
#+END_SRC
** GLSL

#+BEGIN_SRC emacs-lisp
(use-package glsl-mode :ensure t)
#+END_SRC
** Leesp
#+BEGIN_SRC emacs-lisp
;; (use-package geiser
;;   :ensure t
;;   :hook (scheme-mode . geiser-mode)
;; )
; (use-package guile :ensure t)
(use-package paredit
  :ensure t
  :config
  ;; (autoload 'enable-paredit-mode "paredit" "Turn on pseudo-structural editing of Lisp code." t)
  ;; (add-hook 'emacs-lisp-mode-hook       #'enable-paredit-mode)
  ;; (add-hook 'eval-expression-minibuffer-setup-hook #'enable-paredit-mode)
  ;; (add-hook 'ielm-mode-hook             #'enable-paredit-mode)
  ;; (add-hook 'lisp-mode-hook             #'enable-paredit-mode)
  ;; (add-hook 'lisp-interaction-mode-hook #'enable-paredit-mode)
  ;; (add-hook 'scheme-mode-hook           #'enable-paredit-mode)

  (require 'eldoc) ; if not already loaded
  (eldoc-add-command

editor/emacs/config.el
(setq-default indent-tabs-mode nil)

(setq-default tab-width 2)

(setq backup-directory-alist `((".*" . ,temporary-file-directory)))
(setq auto-save-file-name-transforms `((".*" ,temporary-file-directory t)))

(setq gc-cons-threshold 50000000)

;; Enable line highlight mode, because I find it useful.
(global-hl-line-mode t)
;; Column numbers too
(column-number-mode t)
;; Default fill-width
(setq-default fill-column 90)

(use-package diminish
  :ensure t
  :config
  (diminish 'undo-tree-mode "UT")
)

(use-package delight
  :ensure t
)

(setq which-key-enable-extended-define-key t)
(use-package which-key
  :ensure t
  :diminish which-key-mode
  :init
  (setq which-key-separator " ")
  (setq which-key-prefix-prefix "+")
  :config
  (which-key-mode 1)

)

(defun manky/add-leader-prefix-name (k n)
  "Using the current evil-leader key, define a name for which-key to display"
  (which-key-add-key-based-replacements (concat (symbol-value 'evil-leader/leader) " " k) n))

(use-package evil
  :ensure t
  :config

  (use-package evil-leader
    :ensure t
    :config
    (global-evil-leader-mode 1)
    (evil-leader/set-leader "<SPC>"))

  (use-package evil-surround
    :ensure t
    :config (global-evil-surround-mode))

  (use-package evil-escape
    :ensure t
    :diminish evil-escape-mode
    :init (setq-default evil-escape-key-sequence "jk")
    :config (evil-escape-mode 1))

  (use-package evil-nerd-commenter
    :ensure t
    :config
    (manky/add-leader-prefix-name "c" "comments")
    (evil-leader/set-key
      "cl" 'evilnc-comment-or-uncomment-lines
      "cc" 'evilnc-copy-and-comment-lines
      "cp" 'evilnc-comment-or-uncomment-paragraphs
      "cr" 'comment-or-uncomment-region
      "cv" 'evilnc-toggle-invert-comment-line-by-line
      "."  'evilnc-copy-and-comment-operator
      "\\" 'evilnc-comment-operator ; if you prefer backslash key
      )
  )
  ;; (use-package powerline-evil
  ;;   :ensure t
  ;;   :config
  ;;   (powerline-evil-vim-color-theme))
  (evil-mode 1)
)

(use-package avy
  :ensure t
  :config
  (manky/add-leader-prefix-name "j" "avy")
  (evil-leader/set-key
    "j c" 'avy-goto-char
    "j C" 'avy-goto-char-2
    "j t" 'avy-goto-char-timer
    "j l" 'avy-goto-line
    "j w" 'avy-goto-word-1
    "j W" 'avy-goto-word-0
    "j o" 'avy-org-goto-heading-timer
    "j R" 'avy-org-refile-as-child
  )
)

(scroll-bar-mode -1)
(tool-bar-mode   -1)
(tooltip-mode    -1)
(menu-bar-mode   -1)
(setq-default inhibit-startup-screen t)
;; (setq default-frame-alist '((font . "-ADBO-Source Code Pro-normal-normal-normal-*-14-*-*-*-m-0-iso10646-1")))
;; (setq default-frame-alist '((font . "-POOP-Fixedsys Excelsior 3.01-normal-normal-normal-*-16-*-*-*-*-0-iso10646-1")))
(setq default-frame-alist '((font . "-V.R.-PxPlus IBM VGA9-normal-normal-normal-*-15-*-*-*-m-0-iso10646-1")))

(use-package async
  :ensure t
  :config
  (dired-async-mode 1)
)

(use-package popup
  :ensure t
)

(use-package helm
  :ensure t
  :diminish ""
  :bind (
  ("M-x" . helm-M-x)
  ("C-x C-f" . helm-find-files))
  :config
  (helm-mode 1)
)

(use-package helm-ls-git
  :ensure t
  :bind (("C-x C-d" . helm-browse-project))
)

(use-package projectile
  :ensure t
  :after (helm)
  :delight '(:eval (concat " " (projectile-project-name)))
  :init
  (setq projectile-require-project-root nil)
  :config
  (projectile-mode 1)
  ;; (evil-leader/set-key
  ;;   "p" 'projectile-command-map
  ;;   )
)

(use-package helm-projectile
  :ensure t
  :config
  (helm-projectile-on)

  (manky/add-leader-prefix-name "p" "projects")
  (evil-leader/set-key
    "p p" 'helm-projectile-switch-project
    "p f" 'helm-projectile-find-file
    "p b" 'helm-projectile-switch-to-buffer

    ;; helm-projectile-find-file-in-known-projects
    ;; helm-projectile-find-file-dwim
    ;; helm-projectile-find-dir
    ;; helm-projectile-recentf
  )
)

(use-package aggressive-indent
  :ensure t
  :config
  (evil-leader/set-key
    "t a" 'aggressive-indent-mode
  )
)

(use-package rainbow-delimiters
  :ensure t
  ;; There is no global mode, so...
  :hook (prog-mode-hook . rainbow-delimiters-mode)
)

(use-package smartparens
  :ensure t
  :diminish (smartparens-mode . "()")
  :config
  (require 'smartparens-config)
  (smartparens-global-mode t)
  (show-paren-mode t)
)

(use-package magit
  :ensure t
  :diminish magit-auto-revert-mode
  :init
  (manky/add-leader-prefix-name "g" "git")
  (evil-leader/set-key
    "g s" 'magit-status)
)

(use-package direnv
  :ensure t
  :config
  (direnv-mode))

(use-package emmet-mode
  :ensure t
  :init
  (add-hook 'sgml-mode-hook 'emmet-mode) ;; Autostart on markup modes
  (add-hook 'css-mode-hook 'emmet-mode) ;; Emmet has CSS prefix helpers
  (setq emmet-move-cursor-between-quotes t) ;; Move to between the inserted tags

  ;; Not sure if I need this one yet, but I'll know it when I hit it
  ;; (setq emmet-self-closing-tag-style " /") ;; default "/"
  ;; only " /", "/" and "" are valid.
  ;; eg. <meta />, <meta/>, <meta>
)

(use-package nix-mode
  :ensure t
  :mode ("\\.nix\\'" . 'nix-mode)
  :init
  (defun manky/nix-indent ()
    (make-local-variable 'indent-line-function)
    (setq indent-line-function 'nix-indent-line)
    (setq nix-indent-function 'nix-indent-line)
    )

  (add-hook 'nix-mode-hook 'manky/nix-indent)
  )

(use-package nix-sandbox
  :ensure t
  :after nix-mode
  )

(use-package haskell-mode
  :ensure t
  :after flycheck
  :config
  ;; Configure haskell-mode to use cabal new-style builds
  (setq haskell-process-type 'cabal-new-repl)
  ;; Make sure we try to use the current nix env if we have one
  (setq haskell-process-wrapper-function
    (lambda (args) (apply 'nix-shell-command (nix-current-sandbox) args)))

  ;; Disable the haskell-stack-ghc checker
  (add-to-list 'flycheck-disabled-checkers 'haskell-stack-ghc)
  (add-hook 'hack-local-variables-hook #'manky/set-dante-locals nil 'local)

  (add-hook 'haskell-mode-hook
    (lambda ()
            (set (make-local-variable 'company-backends)
                 (append '((company-capf company-dabbrev-code))
                         company-backends))))

)

;; (use-package shm
;;   :load-path "~/repos/structured-haskell-mode/elisp/"
;;   :hook (haskell-mode . structured-haskell-mode)
;;   :init
;;   (setq shm-program-name "/home/manky/repos/structured-haskell-mode/result/bin/structured-haskell-mode")
;;   :config
;;   (haskell-indentation-mode -1)
;; )

(use-package json-mode :ensure t)

(use-package css-mode :ensure t)

(use-package markdown-mode
  :ensure t
)

(use-package glsl-mode
  :ensure t
)

(defun manky/set-dante-locals ()
  (make-local-variable 'dante-project-root)
  (make-local-variable 'dante-repl-command-line)
  (setq dante-project-root (locate-dominating-file buffer-file-name "default..nix"))
  (if dante-target
      (let ((cabal-cmd
             (concat "cabal new-repl " dante-target " --builddir=dist/dante")))
        (setq dante-repl-command-line (list "nix-shell" "--run" cabal-cmd)))
    nil))

(use-package flycheck
  :ensure t
  :init
  (manky/add-leader-prefix-name "t" "toggle")
  (manky/add-leader-prefix-name "e" "fc-errors")
  (evil-leader/set-key
    "t s" 'flycheck-mode
    "e n" 'flycheck-next-error
    "e p" 'flycheck-previous-error
  )
  (setq flycheck-command-wrapper-function
        (lambda (command) (apply 'nix-shell-command (nix-current-sandbox) command))
        flycheck-executable-find
        (lambda (cmd) (nix-executable-find (nix-current-sandbox) cmd)))
)

(use-package dante
  :hook haskell-mode
  :ensure t
  :after haskell-mode
  :commands 'dante-mode
  :init
  (add-hook 'dante-mode-hook
    '(lambda () (flycheck-add-next-checker 'haskell-dante '(warning . haskell-hlint))))

  :config
  (defun manky/dante-insert-type ()
    (interactive)
    (dante-type-at t))

  (evil-leader/set-key-for-mode 'haskell-mode
    "r t" 'manky/dante-insert-type
  )
  (which-key-add-key-based-replacements (concat (symbol-value 'evil-leader/leader) " r t") "insert type")
)

(use-package attrap
  :ensure t
  :init
  (manky/add-leader-prefix-name "r" "refactor")
  (evil-leader/set-key-for-mode 'haskell-mode
    "r f" 'attrap-attrap)
  )

(use-package company
  :ensure t
  :diminish " C"
  :config
  (add-hook 'after-init-hook 'global-company-mode)
)

(use-package smart-mode-line
  :ensure t
  :config
  (add-hook 'after-init-hook 'sml/setup)
)

(manky/add-leader-prefix-name "x" "text") ;; spacemacs muscle memory
(manky/add-leader-prefix-name "f" "file")
(manky/add-leader-prefix-name "b" "buffer")
(which-key-add-key-based-replacements "SPC TAB" "Prev buffer")

;; This is just the beginning
(evil-leader/set-key

"x a r" 'align-regexp

"f s" 'save-buffer

"b d" 'kill-this-buffer
"b b" 'switch-to-buffer
"TAB" 'mode-line-other-buffer

)
;; This is just the end

(use-package org-plus-contrib
  :mode ("\\.org\\'" . org-mode)
  :ensure t
  :pin org
  :config
)
(use-package ox-reveal
  ;; Cloned from github https://github.com/yjwen/org-reveal.git
  :load-path "cloned/org-reveal"
  :config
  (require 'ox-reveal)
)
